### ISSUE TITLE
perhaps zstd-6c61c11  zaps /dev/null on el capitan

### ISSUE THREAD

**Issue by jmscott** [facebook/zstd#316]
I see /dev/null disappear after 'make all && make test' for zstd-master-6c61c11'
on mac osx el capitan 10.11.6.  the compiler version is

       	Apple LLVM version 7.3.0 (clang-703.0.31)
       	Target: x86_64-apple-darwin15.6.0
       	Thread model: posix
       	InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

needless to say, zapping /dev/null cripples the mac.

this error can be easily reproduced by doing

       	cd /usr/local/src
       	unzip ../dist/zstd-master-6c61c11.zip
       	cd zstd-master
       	make all
       	make test

and your /dev/null is gone.

also, to restore /dev/null
       	reboot your mac
       	hold down command-s key combination to enable single user mode
       	at the single user shell, do "shutdown -r now"
       	/dev/null will reappear after this second reboot

**Comment by intdel** [comment]
Are you running make (and make test) in privileged mode (root)?

**Comment by intdel** [comment]
It seems that the issue is in this area (where VOID = /dev/null):

From https://github.com/facebook/zstd/blob/master/tests/Makefile which is called by `make test`

```
valgrindTest: zstd datagen fuzzer fullbench zbufftest
	@echo "\n ---- valgrind tests : memory analyzer ----"
	$(VALGRIND) ./datagen -g50M > $(VOID)
	$(VALGRIND) $(PRGDIR)/zstd ; if [ $$? -eq 0 ] ; then echo "zstd without argument should have failed"; false; fi
	./datagen -g80 | $(VALGRIND) $(PRGDIR)/zstd - -c > $(VOID)
	./datagen -g16KB | $(VALGRIND) $(PRGDIR)/zstd -vf - -o $(VOID)
	./datagen -g2930KB | $(VALGRIND) $(PRGDIR)/zstd -5 -vf - -o tmp
	$(VALGRIND) $(PRGDIR)/zstd -vdf tmp -o $(VOID)
	./datagen -g64MB | $(VALGRIND) $(PRGDIR)/zstd -vf - -o $(VOID)
	@rm tmp
	$(VALGRIND) ./fuzzer -T1mn -t1
	$(VALGRIND) ./fullbench -i1
	$(VALGRIND) ./zbufftest -T1mn
```

I wonder if it overwrites /dev/null (as opposed to write into /dev/null)

**Comment by Cyan4973** [comment]
wow, that's interesting.

Never noticed this behavior, and wouldn't expect it by the way.

So `-o /dev/null` can actually overwrite /dev/null, when given the appropriate rights ?
Good to know, we'll change the tests then

**Comment by intdel** [comment]
I did not test it on mine (as I do not have OS X), but I tried to look at the culprit in the code. So I just suggested what *could* be the issue, not necessarily what the issue *is*! Perhaps I should have phrased my post differently.

**Comment by Cyan4973** [comment]
right, but it makes perfect sense to me.

In anticipation, I updated the "dev" branch
with a change to the mentioned tests to use `-c` instead of `-o`.
It should remove this specific risk of overwriting `/dev/null`.

**Comment by jmscott** [comment]
unfortunately still fails.  the build version is dev 78b102d.

i believe the final error message is different than before.  something
about /dev/null and permissions.  my mac crashes quickly immediately after
running test, so getting a screen snaphot is tricky.


-j

**Comment by Cyan4973** [comment]
Can you try to send anything into `/dev/null` ?
Just to check if this action is enough to crash your system.
`echo bob > /dev/null`

**Comment by jmscott** [comment]
sure ... no problem.

in my experience a non-existent /dev/null will cause bad things to happen
on most unix systems.

you ought to be able to easily reproduce the problem.  my superficial guess
would be a make recipe that is removing $(VOID) as part of cleanup.  that
would also crash a linux server.

-j

**Comment by jmscott** [comment]
box still solvent ...

    dd if=/dev/random of=/dev/null

-j

**Comment by Cyan4973** [comment]
Thanks for the check.

I'm embarrassed, as we apparently have no way of reproducing the bug.
The test script has been and is being played countless number of times on Linux and Mac OS-X machines. But this is the first time this pb occurs.
And unfortunately, without any way to reproduce it, it's very difficult to study and solve the pb ....

**Comment by jmscott** [comment]
check out line 225 in tests/playTests.sh

         $ZSTD -t tmp1 && die "bad file not detected !"

/dev/null exists at line 224 and disappears immediately after the above
invocation.

the error is reproducible via 'make test'.

++ file /dev/null
+ echo 'WTF: 0: /dev/null: character special'
WTF: 0: /dev/null: character special
+ ../programs/zstd -t tmp1
zstd: tmp1: not in zstd format
++ file /dev/null
+ echo 'WTF: 1: /dev/null: cannot open `/dev/null'\'' (No such file or
directory)'
WTF: 1: /dev/null: cannot open `/dev/null' (No such file or directory)

**Comment by intdel** [comment]
John, are you using the default OS X shell or something else? Are you by any chance using iTerminal?

@Cyan4973 
Any idea what "-t" does as I cannot see it in the README.MD?

```
$ECHO "test one file (tmp1.zst) "
./datagen > tmp1
$ZSTD tmp1
$ZSTD -t tmp1.zst
$ZSTD --test tmp1.zst
```

I am not sure how this overwrites (zaps) /dev/null. Did you check if your filesystem is healthy?

**Comment by jmscott** [comment]
i am using iterm2.  build 3.0.7 on a vanilla el capitan built a few months
ago.

file system passes fsck.

check out the final line below.  that "FIO_setRemoveSrcFile(0)" seems
suspicious.

tmonk# pwd
/usr/local/src/zstd-dev

tmonk# find . -name '*.[ch]'|xargs fgrep /dev/null
./programs/fileio.h:#  define nulmark "/dev/null"

tmonk# find . -name '*.[ch]'|xargs fgrep fileio.h
./programs/fileio.c:#include "fileio.h"
./programs/zstdcli.c:#include "fileio.h"

tmonk# find . -name '*.[ch]'|xargs fgrep nulmark
./programs/fileio.c:        if (!g_overwrite && strcmp (dstFileName,
nulmark)) {  /* Check if destination file already exists */
./programs/fileio.c:    if (!strcmp(suffix, stdoutmark) || !strcmp(suffix,
nulmark)) {
./programs/fileio.h:#  define nulmark "nul"
./programs/fileio.h:#  define nulmark "/dev/null"
./programs/zstdcli.c:    if (outFileName && strcmp(outFileName,stdoutmark)
&& strcmp(outFileName,nulmark) && (filenameIdx>1)) {
./programs/zstdcli.c:        if (testmode) { outFileName=nulmark;
FIO_setRemoveSrcFile(0); } /* test mode */


-j

**Comment by Cyan4973** [comment]
`-t` is the same as `--test` : it just checks input validity (and checksum if present).

Technically though, it does send output into `/dev/null`, so there is a risk on this code path too.
`FIO_setRemoveSrcFile(0)` is supposed to mean _don't remove source file_.
Indeed, this code path makes a good suspect.

I also use OS-X el capitan, and routinely run the same test script, without experiencing this specific issue. So I guess there must some other reason involved...

**Comment by jmscott** [comment]
i expect that the single invocation of xstd ought to cause /dev/null to
disappear, once the process environment from the playTests.sh script is
replicated.
gnu is eliminated as the culprit, since i see /dev/null disappear in the
playTests.sh script.

i ought to be able to fully debug tonight (+8 hours).  debugging is tedious
and takes my undivided attention since my mac crashes   i'm sure the bug is
quite simple.

-j

**Comment by Cyan4973** [comment]
Got it.
When doing a test, `zstd` transparently writes the result into `/dev/null`.

When the test fails, typically because it's not a `zstd` conformant file,
`zstd` believes it has created a useless artefact with the name `dstFileName`,
in this case `/dev/null`.
So it `erase()` it.

The call to `erase()` typically fails, which is expected for `/dev/null`.
But in your case, it seems it effectively erases `/dev/null`.
So it's likely a question of rights management.

Additionally, it's also possible to special-case `/dev/null` to not produce an `erase()` invocation in this case.

**Comment by Cyan4973** [comment]
Special case added in `dev` branch

**Comment by intdel** [comment]
I still find it weird that a non-root user can delete /dev/null and cause the system to crash. 

DoS potential...

**Comment by jmscott** [comment]
> I still find it weird that a non-root user can delete /dev/null and cause
the system to crash.

nope, at least not during my tests.

i don't recall seeing a claim that a non-root invocation of zstd zapped
/dev/null.  that would be a unix bug, not an zstd bug.
i only see /dev/null disappear when invoked as root, which implies
*something* is unlink /dev/null from the filesystem.

i will trace the offending invocation of zstd in playTests.sh when i have
dedicated time, hopefully in a day or so, as yann indicated, the error
ought to be easy to reproduce as a single invocation.  i just need a bit
of undivided attention to replicate the shell environment.  each zapping of
/dev/null fully reboots my machine, which makes testing tedious, especially
when paying work is pressing.  give me a day or so.  i'll find the culprit.

-j

**Comment by jmscott** [comment]
good news.   test passes for zstd-dev-a266464.  /dev/null exists after
tests.  any idea what changed?

-j

**Comment by Cyan4973** [comment]
yes : 
https://github.com/facebook/zstd/issues/316#issuecomment-245174884

**Comment by Cyan4973** [comment]
fixed in v1.1.0
